zio-http/jvm/src/test/scala/zio/http/endpoint/AuthSpec.scala

```diff
@@ -44,6 +63,201 @@ object AuthSpec extends ZIOSpecDefault {
+          statusBearer = responseBearer.status
+        } yield assertTrue(
+          statusBasic.isSuccess,
+          bodyBasic == "admin",
+          statusBearer.isSuccess,
+          bodyBearer == "bearer-admin",
+        )
+      },
+      test("Auth from query parameter") {
+        val endpoint = Endpoint(Method.GET / "test")
+          .out[String](MediaType.text.`plain`)
+          .auth(AuthType.Custom(HttpCodec.query("token")))
+        val routes   =
+          Routes(
+            endpoint.implementHandler(handler((_: Unit) => withContext((ctx: AuthContext) => ctx.value))),
+          ) @@ queryParamAuthContext
+        val response = routes.run(
+          Request(
+            method = Method.GET,
+            url = url"/test?token=admin",
+            headers = Headers(
+              Header.Accept(MediaType.text.`plain`),
+            ),
+          ),
+        )
+        for {
+          response <- response
+          body     <- response.body.asString
+        } yield assertTrue(body == "admin")
+      },
+      suite("With server")(
+        test("Auth with context and endpoint client") {
+          val endpoint = Endpoint(Method.GET / "test").out[String](MediaType.text.`plain`).auth(AuthType.Basic)
+          val routes   =
+            Routes(
+              endpoint.implementHandler(handler((_: Unit) => withContext((ctx: AuthContext) => ctx.value))),
+            ) @@ basicAuthContext
+
+          val response = for {
+            client <- ZIO.service[Client]
+            locator    = EndpointLocator.fromURL(url"http://localhost:8080")
+            executor   = EndpointExecutor(client, locator)
+            invocation = endpoint(Header.Authorization.Basic("admin", "admin".reverse))
+            response <- ZIO.scoped(executor(invocation))
+          } yield response
+
+          for {
+            _        <- Server
+              .serve(routes.handleErrorCauseZIO(c => ZIO.logInfoCause("yes!", c).as(Response.text(""))))
+              .forkDaemon
+              .catchAllCause(c => ZIO.logInfoCause(c)) <* ZIO.sleep(1.seconds)
+            response <- response
+          } yield assertTrue(response == "admin")
+        },
+        test("Auth basic or bearer with context and endpoint client") {
+          val endpoint =
+            Endpoint(Method.GET / "test" / "multiAuth")
+              .out[String](MediaType.text.`plain`)
+              .auth(AuthType.Basic | AuthType.Bearer)
+          val routes   =
+            Routes(
+              endpoint.implementHandler(handler((_: Unit) => withContext((ctx: AuthContext) => ctx.value))),
+            ) @@ basicOrBearerAuthContext
+
+          val responseBasic = for {
+            client <- ZIO.service[Client]
+            locator    = EndpointLocator.fromURL(url"http://localhost:8080")
+            executor   = EndpointExecutor(client, locator)
+            invocation = endpoint(Left(Header.Authorization.Basic("admin", "admin".reverse)))
```

Comment:
Out of curiosity, would it possible to replace it with some kind of tagged union to avoid nested `Either` types?

Scala 3 has union type and `scala.reflect.TypeTest` but I'm not sure of the Scala 2 equivalent.

Reply:
We can add a Scala 3 specific variant of the or. Like in zio-schema
https://github.com/zio/zio-schema/blob/5a276f7417b3039f46da7c85ca41e00c4f993be0/zio-schema/shared/src/main/scala-3/zio/schema/package.scala#L15